### PR TITLE
add current version of safety oracle

### PR DESCRIPTION
**Pull request by naterush**
TODO:
- [ ] formalize induction on minimal transitions
- [ ] change language from safety -> decisions to be consistent with main paper
- [ ] major cleanup, refactoring, and sanity checking :)

### CODE REVIEW

File: safety-oracle.tex [hunk 1]
```diff
@@ -0,0 +1,1552 @@
...
+
+This function is not defined when its argument is not singleton (i.e. when is multi-element or the empty set).
+
+\subsection{Validator Cliques}
+
+A clique of validators for some proposition $p$ can be thought of as a set of validators who pairwise see eachother agreeing with $p$, where there are also no later messages from these validators that are disagreeing with $p$. A clique can have mulitple layers, although the minimum viable safety oracle will only be $1-layer$ deep. \\
+
+\begin{defn}[Clique with $n$ layers]
+\begin{align}
+Clique&: \mathbb{N}^+ \times \mathcal{P}(\mathcal{V}) \times P_{\mathcal{C}} \times \Sigma \to \{True, False\} \\
```
Comment (nrryuya) [inline safety-oracle.tex:172]: `!`s are missing just before `L^H_J` and `L^H_M`? 

```
Clique&: \mathbb{N}^+ \times \mathcal{P}(\mathcal{V}) \times P_{\mathcal{C}} \times \Sigma \to \{True, False\} \\
Clique(1, V, p, \sigma) &:\Leftrightarrow \forall v \in V, V \subseteq Agreeing(p, !L^H_J(\sigma)(v)) \\
&~~~~\land \forall v' \in V, Later\_Disagreeing(p, !L^H_M(!L^H_J(\sigma)(v))(v'), v', \sigma) = \emptyset \\
Clique(n, V, p, \sigma) &:\Leftrightarrow \forall v \in V, V \subseteq Agreeing(p, !L^H_J(\sigma)(v)) \\
&~~~~\land \forall v \in V, Clique\_Cond(n - 1, V, p, !L^H_J(\sigma)(v))
```

File: safety-oracle.tex [hunk 2]
```diff
@@ -0,0 +1,1552 @@
...
+$$
+Weight: \mathcal{P}(\mathcal{V}) \to \mathbb{R}_+
+$$
+$$
+Weight(V) :\Leftrightarrow \sum_{v \in V} \mathcal{W}(v)
+$$
+\end{defn}
+
+
+\begin{defn}[Non-equivocating majority driven property]
```
Comment (nrryuya) [inline safety-oracle.tex:116]: > `Non-equivocating majority driven property`

It should be `Majority validators`, `A validator set is a majority` or something?

File: safety-oracle.tex [hunk 3]
```diff
@@ -0,0 +1,1552 @@
...
+\implies&\{m \in \sigma: Sender(m) = \overline{v} \land Later(m, \overline{m})\} = \emptyset \land \neg Later(\overline{m}, \overline{m}) \\
+\implies&\{m \in \sigma: Sender(m) = \overline{v} \land Later(m, \overline{m})\} = \emptyset \land \{m \in \{\overline{m}\}: Sender(m) = \overline{v} \land Later(m, \overline{m})\} = \emptyset \\
+\implies&\{m \in \sigma: Sender(m) = \overline{v} \land Later(m, \overline{m})\} \cup \{m \in \{\overline{m}\}: Sender(m) = \overline{v} \land Later(m, \overline{m})\} = \emptyset \\
+\implies&\{m \in \sigma \cup \{\overline{m}\}: Sender(m) = \overline{v} \land Later(m, \overline{m})\} = \emptyset \\
+\implies&\{m \in \sigma': Sender(m) = \overline{v} \land Later(m, \overline{m})\} = \emptyset \\
+\implies&Later\_From(\overline{m}, \overline{v}, \sigma') = \emptyset \\
+\end{align}
+We will now be able to show that $\overline{m}$ is a latest honest message. Note that if a validator is not equivocating in $\sigma'$, the validators latest honest message is equal to the validators latest messages, by the defintion of latest honest message.
```
Comment (nrryuya) [inline safety-oracle.tex:583]: `defintion`
Typo ;)

File: safety-oracle.tex [hunk 4]
```diff
@@ -0,0 +1,1552 @@
...
+$\forall (\sigma, \sigma') \in Minimal_t$, letting $\overline{m} =~!\sigma'\setminus\sigma$, and $\overline{v} = Sender(\overline{m})$
+$$
+\forall v \in \mathcal{V} \setminus \{\overline{v}\}, L^H_M(\sigma)(v) = L^H_M(\sigma')(v)
+$$
+\end{lemma}
+
+\begin{proof}
+Due to the definition of $L^H_M$, there are two cases: when $v \in E(\sigma)$ and otherwise:
+
+\[ L^H_M(\sigma))(v) = \left\{
```
Comment (nrryuya) [inline safety-oracle.tex:268]: > L^H_M(\sigma))(v)

The second `)` is redundant.

File: safety-oracle.tex [hunk 5]
```diff
@@ -0,0 +1,1552 @@
...
+\end{align}
+Thus, this is clearly a contradiction, as the latest message in a state cannot have any later messages. Thus, we know that:
+$$
+\neg (\exists m \in Justification(\overline{m}): Sender(m) = \overline{v} \land Later(m, !L^H_M(\sigma)(\overline{v})))
+$$
+which is enough to complete the proof.
+\begin{align}
+          &\neg (\exists m \in Justification(\overline{m}): Sender(m) = \overline{v} \land Later(m, !L^H_M(\sigma)(\overline{v}))) \\
+  \implies&\nexists m \in Justification(\overline{m}): Sender(m) = \overline{v} \land Later(m, !L^H_M(\sigma)(\overline{v})) \\
+  \implies&\{m \in Just
```
Comment (nrryuya) [inline safety-oracle.tex:1343]: The function signature is wrong.

Comment (nrryuya) [reply safety-oracle.tex:1343]: and the order of params `σ` and `p` of `Clique` is reversed from the definition 7.16?

File: safety-oracle.tex [hunk 6]
```diff
@@ -0,0 +1,1552 @@
...
+\end{defn}
+
+
+\subsection{Cliques Survive Messages from Validators Outside Clique}
+
+First, we define minimal transitions:
+
+\begin{defn}[Minimal transitions]
+$$
+Minimal_t = \{ (\sigma, \sigma') \in \Sigma_t^2: \sigma \to \sigma' \land \nexists \sigma'' . \sigma \to \sigma'' \to \sigma' \land \sigma'' \neq \sigma \land \sigma'' \neq \sigma' \}
```
Comment (nrryuya) [inline safety-oracle.tex:187]: I guess we also need `σ ≠ σ'` for a minimal transition?
If `σ = σ'`,  `σ \ σ' = ∅` and this is not good in the later lemmas.

File: safety-oracle.tex [hunk 7]
```diff
@@ -0,0 +1,1552 @@
...
+\implies&\forall m \in \sigma: Sender(m) = \overline{v}, m \in Justification(\overline{m})\\
+\end{align}
+\end{proof}
+\fi
+
+
+\begin{lemma}[$\overline{m}$ is $\overline{v}$'s latest message in $\sigma$']
+$\forall (\sigma, \sigma') \in Minimal_t$, letting $\overline{m} =~!\sigma'\setminus\sigma$ and $\overline{v} = Sender(\overline{m})$,
+$$
+\overline{v} \notin E(\sigma') \implies \overline{m} = !L^H_M(\sigma')(\overline{v})))
```
Comment (nrryuya) [inline safety-oracle.tex:560]: > !L^H_M(\sigma')(\overline{v})))

`))` is redundant.

File: safety-oracle.tex [hunk 8]
```diff
@@ -0,0 +1,1552 @@
...
+\end{defn}
+
+
+The validators whose latest estimates agree with some proposition $p$ in some protocol state $\sigma$ is denoted formally as follows
+\begin{defn}[Validators agreeing with $p$ in $\sigma$]
+$$
+Agreeing: P_{\mathcal{C}} \times \Sigma \to \mathcal{P}(\mathcal{V})
+$$
+$$
+Agreeing(p, \sigma) :\Leftrightarrow \{v \in \mathcal{V} : \forall c \in L^H_E(\sigma)(v), p(c)\}
```
Comment (nrryuya) [inline safety-oracle.tex:90]: In HOL, `∀ a ∈ ∅. P` is True. Therefore, `v` s.t. `L^H_E(\sigma)(v)` is empty is included in `agreeing`.
I think it should be re-defined to remove validators which are not observed or equivocating like this (Isabelle code):
```
fun observed_non_equivocating_validators :: "state ⇒ validator set"
  where
    "observed_non_equivocating_validators σ = observed σ - equivocating_validators σ"

fun  agreeing_validators :: "(consensus_value_property * state) ⇒ validator set"
  where
    "agreeing_validators (p, σ) = {v ∈ observed_non_equivocating_validators σ. ∀ c ∈ latest_estimates_from_non_equivocating_validators σ v. p c}"
```

File: safety-oracle.tex (hunk 5 above)
Comment (nrryuya) [inline safety-oracle.tex:1526]: `forall` → `\forall`

File: safety-oracle.tex [hunk 9]
```diff
@@ -0,0 +1,1552 @@
...
+
+As above, $L^H_E(\sigma)(v) = \emptyset$ for validators $v$ who are equivocating or missing in $\sigma$.
+
+
+\begin{defn}[Latest justifications from non-Equivocating validators]
+$$
+L^H_J:\Sigma \to (\mathcal{V} \to \mathcal{P}(\Sigma))
+$$
+$$
+L^H_J(\sigma)(v) = \{Justification(m) : m \in L^H_M(\sigma)(v)\}
```
Comment (afck) [inline safety-oracle.tex:79]: Should this be the _union_ (or `!`) of that set, i.e. just `Justification(m)` for the unique `m \in L^H_M(\sigma)(v)`?

File: safety-oracle.tex [hunk 10]
```diff
@@ -0,0 +1,1552 @@
...
+
+First, we define minimal transitions:
+
+\begin{defn}[Minimal transitions]
+$$
+Minimal_t = \{ (\sigma, \sigma') \in \Sigma_t^2: \sigma \to \sigma' \land \nexists \sigma'' . \sigma \to \sigma'' \to \sigma' \land \sigma'' \neq \sigma \land \sigma'' \neq \sigma' \}
+$$
+\end{defn}
+
+A minimal transition corresponds to recieving a single new message with justification drawn from the initial protocol state. We will show that any minimal transition from a validator not in the clique will not affect the clique.
```
Comment (afck) [inline safety-oracle.tex:191]: Typo: "receiving"

File: safety-oracle.tex [hunk 11]
```diff
@@ -0,0 +1,1552 @@
...
+$$
+!(A) = a \in A : \forall b \in A, b = a
+$$
+\end{defn}
+
+This function is not defined when its argument is not singleton (i.e. when is multi-element or the empty set).
+
+\subsection{Validator Cliques}
+
+A clique of validators for some proposition $p$ can be thought of as a set of validators who pairwise see eachother agreeing with $p$, where there are also no later messages from these validators that are disagreeing with $p$. A clique can have mulitple layers, although the minimum viable safety oracle will only be $1-layer$ deep. \\
```
Comment (0xjac) [inline safety-oracle.tex:168]: - s/eachother/each other/
- s/mulitple/multiple/

File: safety-oracle.tex [hunk 12]
```diff
@@ -0,0 +1,1552 @@
...
+
+\begin{lemma}[Minimal transitions do not change equivocation status for any non-sender]
+$\forall (\sigma, \sigma') \in Minimal_t$, letting $\overline{m} =~!\sigma'\setminus\sigma$, and $\overline{v} = Sender(\overline{m})$,
+$$
+\forall v \in \mathcal{V} \setminus \{\overline{v}\}, (v \in E(\sigma) \iff v \in E(\sigma'))
+$$
+\end{lemma}
+
+\begin{proof}
+The forward direction, $v \in E(\sigma) \implies v \in E(\sigma')$, follows immediatly as equivocation is monotonic,
```
Comment (0xjac) [inline safety-oracle.tex:225]: s/immediatly/immediately/

**Comment by goral09** [comment]
What's the status of this? This PR hasn't been given love in a while and it's pretty important.

**Pull request closed without merging**